tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,166 @@
+use futures::prelude::*;
+use serde::{Deserialize, Serialize};
+
+use tokio::net::TcpStream;
+
+/// Event Listener over webocket. https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(address: net::Address) -> Result<EventListener, RPCError> {
+        let (host, port) = match address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(RPCError::invalid_params(&format!(
+                    "invalid RPC address: {:?}",
+                    other
+                )));
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: &str) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.to_owned()).into_json(),
+            ))
+            .await?;
+        Ok(())
+    }
+
+    /// Get the next event from the websocket
+    pub async fn get_event(&mut self) -> Result<Event, RPCError> {
+        let msg = self
+            .socket
+            .next()
+            .await
+            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+
+        match serde_json::from_str::<JSONRPC>(&msg.to_string()) {
+            Ok(data) => Ok(Event::JsonRPCTransactionResult {
+                data: Box::new(data),
+            }),
+
+            Err(_) => match msg.to_string().parse::<serde_json::Value>() {
+                Ok(data) => Ok(Event::GenericJSONEvent { data }),
+                Err(_) => Ok(Event::GenericStringEvent {
+                    data: msg.to_string(),
+                }),
+            },
+        }
+    }
+}
+
+//TODO more event types
+/// The Event enum is typed events emitted by the Websockets
+#[derive(Debug, Clone)]
+pub enum Event {
+    /// The result of the ABCI app processing a transaction, serialized as JSON RPC response
+    JsonRPCTransactionResult {
+        /// the tx result data
+        data: Box<JSONRPC>,
+    },
+
+    ///Generic event containing json data
+    GenericJSONEvent {
+        /// generic event json data
+        data: serde_json::Value,
+    },
+    ///Generic String Event
+    GenericStringEvent {
+        /// generic string data
+        data: String,
+    },
+}
+
+/// Standard JSON RPC Wrapper
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct JSONRPC {
+    jsonrpc: String,
+    id: String,
+    result: RPCResult,
+}
```

Comment:
This kinda duplicates the existing JSONRPC wrapper:
https://github.com/informalsystems/tendermint-rs/blob/9fc495a88803076b177ec90b8bb0a0b6d3102f47/tendermint/src/rpc/response.rs#L23-L37

Reply:
This was addressed in e1549e83e71bee7d846011aef483b895f3a2d403